ENH: add render_links for Styler.to_html formatting

attack68 opened the pull request:
This came about as a requirement for converting DataFrame.to_html into Styler.to_html, citing @jorisvandenbossche in favour of the functionality.

Note that the `DataFrame.to_html` `render_links` argument is rather limited to just detecting only a URL in a cell.

The function added for Styler here is a more general pattern search, and it allows the result to be viewed in Jupyter Notebook directly, which the precursor does not.


![Screen Shot 2021-12-24 at 22 20 45](https://user-images.githubusercontent.com/24256554/147372493-5458f18a-4cf3-417a-86a1-7750e430b145.png)

attack68 committed:
render links

attack68 committed:
tests

attack68 committed:
whats new

mroeschke commented on pandas/io/formats/style_render.py:
```diff
@@ -1391,13 +1406,23 @@ def _str_escape(x, escape):
+    """uses regex to detect a common URL pattern and converts to HTML <a> tag"""
+    if isinstance(x, str):
+        href = '<a href="{0}" target="_blank">{0}</a>'
+        pat = r"(https?:\/\/|ftp:\/\/|www.)[\w/\-?=%.]+\.[\w/\-&?=%.]+"
```
Is there a way to somehow leverage `pandas.io.common.is_url`?

attack68 replied:
I looked but I don't believe this finds URLs within a random string. Is why I used `re`. Maybe wrong.

mroeschke replied:
Hmm maybe you can leverage `is_url` to avoid false positives from the regex like

```
url = re.search(pat, x).group()
if pandas.io.common.is_url(url):
     ...
```

attack68 replied:
Thats a nice idea, because the regex will permit some invalid urls, like `http://.1.2`, but the `is_url` function doesn't really help here since it detects that as a `True` url anyway, and it is more restrictive since it requires a `scheme` so `www.1.2` is apparently an invalid URL, according to `is_url`, since the scheme is not specified (a browser will default to using `https://` I believe). 

Happy if someone with more experience can improve this basic pattern search and url check, but I think as it stands is a good enough start

mroeschke replied:
Gotcha. Okay this regex approach is good for now.

mroeschke commented on pandas/tests/io/formats/style/test_html.py:
```diff
@@ -764,3 +764,33 @@ def test_hiding_index_columns_multiindex_trimming():
+        ("https scheme: https://www.web.com", True, "https://www.web.com"),
+        ("ftp scheme: ftp://www.web", True, "ftp://www.web"),
+        ("subdirectories: www.web.com/directory", True, "www.web.com/directory"),
+        ("Multiple domains: www.1.2.3.4", True, "www.1.2.3.4"),
```
What if there are multiple valid links in the text?

attack68 replied:
they each get rendered individually. see added test

attack68 committed:
multipl links test

attack68 committed:
add LaTeX. chg to `hyperlinks` instead of `render_links`

attack68 commented:
Please note I added the LaTeX equivalent, so the signature is now consistent with the `escape` argument (in `{"html", "latex"}`.

Also note I changed the argument name to `hyperlinks` from `render_links`. Comments welcome.

attack68 committed:
whats new update

attack68 committed:
extend docs

jreback commented on pandas/io/formats/style_render.py:
```diff
@@ -798,6 +798,7 @@ def format(
         decimal: str = ".",
         thousands: str | None = None,
         escape: str | None = None,
+        hyperlinks: str | None = None,
```
is there a reason we prefer this to `render_links`?

attack68 replied:
My arguments: `hyperlink` is a bit more specific whereas `link` can be a general term in certain contexts. And, technically, everything is rendered so why not add `render` to every kwarg. And rendering is somewhat of an esoteric concept outside of graphics processing.

jreback replied:
ok fair

jreback commented:
thanks @attack68

jreback commented:
i was not sure if there was an actual issue for this (aside from the discussion), if so can you link.
